### PR TITLE
Issues/#1489 proxy error fix

### PR DESCRIPTION
**Pull request by alexlehm**
fix for #1489, pass exception to channel handler in ChannelInboundHandlerAdapter
add unit tests for the issue

### CODE REVIEW

File: src/test/java/io/vertx/test/core/ConnectHttpProxy.java [hunk 1]
```diff
@@ -64,7 +66,10 @@ public void start(Vertx vertx, Handler<Void> finishedHandler) {
           return;
         }
       }
-      if (method != HttpMethod.CONNECT || !uri.contains(":")) {
+      if (error != 0) {
+        log.info("sending error " + error);
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/ConnectHttpProxy.java]: please remove this log

File: src/test/java/io/vertx/test/core/ProxyErrorTest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            .setHost("localhost")
+            .setPort(proxy.getPort()));
+    HttpClient client = vertx.createHttpClient(options);
+
+    client.getAbs("https://localhost/", resp -> {
+      log.info("this request is supposed to fail");
+      fail();
+    })
+    .exceptionHandler(e -> {
+      log.warn("Exception", e);
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/ProxyErrorTest.java]: please no logging

File: src/test/java/io/vertx/test/core/ProxyErrorTest.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      latch.countDown();
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/ProxyErrorTest.java]: use testComplete() like other tests

File: src/test/java/io/vertx/test/core/ProxyErrorTest.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    })
+    .end();
+
+    latch.await();
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/ProxyErrorTest.java]: use just await()

File: src/test/java/io/vertx/test/core/ProxyErrorTest.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    final HttpClientOptions options = new HttpClientOptions()
+        .setProxyOptions(new ProxyOptions()
+            .setType(ProxyType.HTTP)
+            .setHost("localhost")
+            .setPort(proxy.getPort()));
+    HttpClient client = vertx.createHttpClient(options);
+
+    client.getAbs("https://localhost/", resp -> {
+      log.info("this request is supposed to fail");
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/ProxyErrorTest.java]: please no logging - a comment rather

File: src/test/java/io/vertx/test/core/ProxyErrorTest.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            .setHost("localhost")
+            .setPort(proxy.getPort()));
+    HttpClient client = vertx.createHttpClient(options);
+
+    client.getAbs("https://unknown.hostname/", resp -> {
+      log.info("this request is supposed to fail");
+      fail();
+    })
+    .exceptionHandler(e -> {
+      log.warn("Exception", e);
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/ProxyErrorTest.java]: please no logging

**Comment by alexlehm** [comment]
@vietj remove the log and change to use test methods without own latch

File: src/test/java/io/vertx/test/core/ProxyErrorTest.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    final HttpClientOptions options = new HttpClientOptions()
+        .setProxyOptions(new ProxyOptions()
+            .setType(ProxyType.HTTP)
+            .setHost("localhost")
+            .setPort(proxy.getPort()));
+    HttpClient client = vertx.createHttpClient(options);
+
+    client.getAbs("https://unknown.hostname/", resp -> {
+      // request is supposed to fail
+      fail();
```
Comment (vietj) [inline src/test/java/io/vertx/test/core/ProxyErrorTest.java]: should be testComplete()

**Pull request merged**